man/gde2-desktop-item-edit.1
.\" Man page for gde2-desktop-item-edit
.TH GDE2-DESKTOP-ITEM-EDIT 1 "4 February 2014" "GDE2 Desktop Environment"
.\" Please adjust this date when revising the manpage.
.\"
.SH "NAME"
gde2-desktop-item-edit \- A small tool to edit .desktop files.
.SH "SYNOPSIS"
.B gde2-desktop-item-edit [OPTIONS] [FILE]
.SH "DESCRIPTION"
.PP 
Distributed with gde2-panel, \fBgde2-desktop-item-edit\fR allows you to edit .desktop files. 
.SH "OPTIONS"
.TP
\fB\-\-create-new\fR 
Create a new file in the given directory. 
.TP
\fB\-\-display=DISPLAY\fR
X display to use.
.TP
\fB\-?, \-h, \-\-help\fR
Print standard command line options.
.TP
\fB\-\-help\-all\fR
Print all command line options.
.P
This program also accepts the standard GTK options.
.SH "EXAMPLES"
\fBgde2-desktop-item-edit \-\-create\-new ~/Desktop/Icon.desktop\fR
.RS 4
Open a \fBgde2-desktop-item-edit\fR dialog to create a launcher icon on the desktop named "Icon".
.SH "BUGS"
.SS Should you encounter any bugs, they may be reported at: 
http://github.com/gde2-desktop/gde2-panel/issues
.SH "AUTHORS"
.SS This Man Page has been updated/re-written for the GDE2 Desktop Environment by:
Adam Erdman <hekel@archlinux.info> (2014)
.SH "HISTORY"
This manual page was originally written by Christian Marillat <marillat@debian.org> for the Debian GNU/Linux system. (2006)
.SH "SEE ALSO"
.SS Further information may also be available at: http://wiki.gde2-desktop.org/docs


NEWS
1.14.2
  * Some fixes for opensuse.layout
  * Some more fixes and cleanups
  * Translations update
  * Make distcheck use currently selected GTK+ version

1.14.1
  * Fix missing focus for panel widgets (needed for Orca to work)
  * Some more fixes and cleanups
  * Translations update

1.14.0
  * Drop Gde2Conf migration script
  * Add separate "Delete" button in reload applet dialog, and
    make "Don't Reload" button do what it says (it won't delete
    applet permanently now)
  * Make max number of bookmarks in menu (before submenu is created)
    configurable via GSettings key
  * Make icon size of menu items configurable via GSettings key
  * Change default menubar icon size to 22px
  * Drop "Ubuntu hacks" (--enable-ubuntu build option) as Ubuntu
    dropped that stuff long ago
  * GTK+3: add several style classes
  * GTK+3: add several css names for 3.20
  * GTK+3: fix rendering of custom backgrounds with 3.18 and newer
  * GTK+3: fix a lot of deprecations
  * Some more fixes and cleanups
  * Translations update
  * Add Changelog generation

1.12.2
  * Drawer applet: fix occasional stretching of drawer's panel
    over the entire screen
  * Drawer applet: fix occasional crash on applet removing
  * GTK+3: fix infinite loop with some themes
  * Some more fixes

1.12.1
  * GTK+3: fix background in some themes
  * Some more fixes

1.12.0
  * Require xrandr 1.3
  * Add option to enable/disable week numbers in calendar
  * Fix positions of locked applets on screen resolution change
  * Retrieve strings directly from gschema (requires intltool 0.50.1)
  * Switch to org.gnome.SessionManager (used by action buttons)
  * GTK+3: add a style class for applets to follow panel-menu-bar
  * Multiple GTK+3 fixes
  * Some more fixes and cleanups

1.10.1
  * Fish and clock applets: help fixes
  * Drawer applet: fix opening after adding applets to it
  * Force Quit applet: fix desktop freezing with GTK+3
  * Some more GTK+3 fixes
  * Some code cleanup
  * Translations updates

1.10.0
  * Gtk3, Fix deprecated use of gdk pointer api
  * Gtk3, Port all GtkStyle to GtkStyleContext
  * Gtk3, Port gtk_rc_parse_string to GtkCssProvider
  * Gtk3, Several small fixes
  * Several more fixes from monsta
  * GDK_KEY_* is also in Gtk+2.24
  * showdesktop: Fix the size of the icon
  * clock: Check allocated size before creating the shadow pixbuf
  * applet: Disconnect show and deactivate menu signals before destroying the applet menu
  * Add a gtkimage with the current pixbuf to button-widget
  * Fix implicit declaration of function of gdk_x11_set_sm_client_id
  * panel-addto: Remove the vbox as it is redundant.
  * na: Fix handling of SYSTEM_TRAY_CANCEL_MESSAGE
  * Use gtk icon cache instead of custom one
  * Several small fixes by monsta

1.8.0
  * Added support for multiple layouts, stored in $(datadir)/gde2-panel/layouts
  * Added --layout parameter
  * Add hook to support run dialog and main menu opening with metacity
  * Switch to libwnck
  * Use yelp-tools instead of gde2-doc-utils

1.6.0
  * add --run-dialog option to open run dialog
  * migration to gsettings
  * panel-menu-button: supports changed notify for new "has-arrow" key
  * clock applet: removed evolution support
  * default layout: default panel layout is loaded from
    $(datadir)/gde2-panel/panel-default-layout.gde2. You can ovverride it by
    providing $(datadir)/gde2-panel/panel-default-layout.dist.


gde2-panel/gde2-desktop-item-edit.c
#include "config.h"

#include <glib/gi18n.h>
#include <gio/gio.h>
#include <gtk/gtk.h>

#include <libpanel-util/panel-error.h>
#include <libpanel-util/panel-keyfile.h>

#include "panel-ditem-editor.h"
#include "panel-icon-names.h"
#include "panel-util.h"

/* FIXME Symbols needed by panel-util.c - sucky */
#include "applet.h"
GSList *gde2_panel_applet_list_applets (void) { return NULL; }
#include "panel-config-global.h"
gboolean panel_global_config_get_tooltips_enabled (void) { return FALSE; }
#include "panel-lockdown.h"
gboolean panel_lockdown_get_disable_lock_screen (void) { return FALSE; }

static int dialogs = 0;
static gboolean create_new = FALSE;
static char **desktops = NULL;

static GOptionEntry options[] = {
	{ "create-new", 0, 0, G_OPTION_ARG_NONE, &create_new, N_("Create new file in the given directory"), NULL },
	{ G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &desktops, NULL, N_("[FILE...]") },
	{ NULL }
};

static void
dialog_destroyed (GtkWidget *dialog, gpointer data)
{
	dialogs --;

	if (dialogs <= 0)
		gtk_main_quit ();
}

static void
validate_for_filename (char *file)
{
	char *ptr;

	g_return_if_fail (file != NULL);
	
	ptr = file;
	while (*ptr != '\0') {
		if (*ptr == '/')
			*ptr = '_';
		ptr++;
	}
}

static char *
find_uri_on_save (PanelDItemEditor *dialog,
		  gpointer          data)
{
	GKeyFile *keyfile;
	char     *name;
	char     *filename;
	char     *uri;
	char     *dir;

	keyfile = panel_ditem_editor_get_key_file (dialog);
	name = panel_key_file_get_string (keyfile, "Name");

	validate_for_filename (name);
	filename = g_filename_from_utf8 (name, -1, NULL, NULL, NULL);
	g_free (name);

	if (filename == NULL)
		filename = g_strdup ("foo");

	dir = g_object_get_data (G_OBJECT (dialog), "dir");
	uri = panel_make_unique_desktop_path_from_name (dir, filename);

	g_free (filename);

	return uri;
}

static void
error_reported (GtkWidget  *dialog,
		const char *primary,
		const char *secondary,
		gpointer    data)
{
	panel_error_dialog (GTK_WINDOW (dialog), NULL,
			    "error_editing_launcher", TRUE,
			    primary, secondary);
}

int
main (int argc, char * argv[])
{
	GError *error = NULL;
	int i;

	bindtextdomain (GETTEXT_PACKAGE, GDE2LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	if (!gtk_init_with_args (&argc, &argv,
	                         _("- Edit .desktop files"),
	                         options,
	                         GETTEXT_PACKAGE,
	                         &error)) {
		g_printerr ("%s\n", error->message);
		g_error_free (error);
		return 1;
	}

	gtk_window_set_default_icon_name (PANEL_ICON_LAUNCHER);

	if (desktops == NULL ||
	    desktops[0] == NULL) {
		g_printerr ("gde2-desktop-item-edit: no file to edit\n");
		return 0;
	}

	for (i = 0; desktops[i] != NULL; i++) {
		GFile     *file;
		GFileInfo *info;
		GFileType  type;
		char      *uri;
		char      *path;
		GtkWidget *dlg = NULL;

		file = g_file_new_for_commandline_arg (desktops[i]);
		uri  = g_file_get_uri (file);
		path = g_file_get_path (file);
		info = g_file_query_info (file, "standard::type",
					  G_FILE_QUERY_INFO_NONE, NULL, NULL);
		g_object_unref (file);

		if (info) {
			type = g_file_info_get_file_type (info);

			if (type == G_FILE_TYPE_DIRECTORY && create_new) {

				dlg = panel_ditem_editor_new (NULL, NULL, NULL,
							     _("Create Launcher"));
				g_object_set_data_full (G_OBJECT (dlg), "dir",
							g_strdup (path),
							(GDestroyNotify)g_free);

				panel_ditem_register_save_uri_func (PANEL_DITEM_EDITOR (dlg),
								    find_uri_on_save,
								    NULL);

			} else if (type == G_FILE_TYPE_DIRECTORY) {
				/* Rerun this iteration with the .directory
				 * file
				 * Note: No need to free, for one we can't free
				 * an individual member of desktops and
				 * secondly we will soon exit */
			        desktops[i] = g_build_path ("/", uri,
							    ".directory", NULL);
				i--;
			} else if (type == G_FILE_TYPE_REGULAR
				   && g_str_has_suffix (desktops [i],
					   		".directory")
				   && !create_new) {
				dlg = panel_ditem_editor_new_directory (NULL,
									NULL,
									uri,
									_("Directory Properties"));
			} else if (type == G_FILE_TYPE_REGULAR
				   && g_str_has_suffix (desktops [i],
					   		".desktop")
				   && !create_new) {
				dlg = panel_ditem_editor_new (NULL, NULL, uri,
							      _("Launcher Properties"));
			} else if (type == G_FILE_TYPE_REGULAR
				   && create_new) {
				g_printerr ("gde2-desktop-item-edit: %s "
					    "already exists\n", uri);
			} else {
				g_printerr ("gde2-desktop-item-edit: %s "
					    "does not look like a desktop "
					    "item\n", uri);
			}

			g_object_unref (info);

		} else if (g_str_has_suffix (desktops [i], ".directory")) {
			/* a non-existant file.  Well we can still edit that
			 * sort of.  We will just create it new */
			dlg = panel_ditem_editor_new_directory (NULL, NULL, uri,
								_("Directory Properties"));

		} else if (g_str_has_suffix (desktops [i], ".desktop")) {
			/* a non-existant file.  Well we can still edit that
			 * sort of.  We will just create it new */
			dlg = panel_ditem_editor_new (NULL, NULL, uri,
						      _("Create Launcher"));

		} else {
			g_printerr ("gde2-desktop-item-edit: %s does not "
				    "have a .desktop or .directory "
				    "suffix\n", uri);
		}

		if (dlg != NULL) {
			dialogs ++;
			g_signal_connect (G_OBJECT (dlg), "destroy",
					  G_CALLBACK (dialog_destroyed), NULL);
			g_signal_connect (G_OBJECT (dlg), "error_reported",
					  G_CALLBACK (error_reported), NULL);
			gtk_widget_show (dlg);
		}

		g_free (uri);
		g_free (path);
	}

	if (dialogs > 0)
		gtk_main ();

        return 0;
}
